haystack/document_stores/types/filter_policy.py

```diff
@@ -52,10 +83,48 @@ def apply_filter_policy(
           values from the runtime filters will overwrite those from the initial filters.
     :param init_filters: The initial filters set during the initialization of the relevant retriever.
     :param runtime_filters: The filters provided at runtime, usually during a query operation execution. These filters
-                            can change for each query/retreiver run invocation.
+                            can change for each query/retriever run invocation.
+    :param default_logical_operator: The default logical operator to use when merging filters (non-legacy filters only).
     :returns: A dictionary containing the resulting filters based on the provided policy.
     """
     if filter_policy == FilterPolicy.MERGE and runtime_filters:
-        return {**(init_filters or {}), **runtime_filters}
-    else:
-        return runtime_filters or init_filters
+        # legacy filters merge handling
+        if is_legacy(runtime_filters):
+            return {**(init_filters or {}), **runtime_filters}
+        elif init_filters is not None:
+            # here we merge new filters
+            def merge_comparison_filters(
+                filter1: Dict[str, Any], filter2: Dict[str, Any], logical_op: str
+            ) -> Dict[str, Any]:
+                if filter1["field"] == filter2["field"]:
+                    # When fields are the same, use the runtime filter (filter2)
+                    return filter2
+                return {"operator": logical_op, "conditions": [filter1, filter2]}
+
+            def merge_comparison_and_logical(
+                comparison: Dict[str, Any], logical: Dict[str, Any], logical_op: str
+            ) -> Dict[str, Any]:
+                if logical["operator"] == logical_op:
+                    logical["conditions"].append(comparison)
+                    return logical
+                else:
+                    return {"operator": logical_op, "conditions": [comparison, logical]}
+
+            def merge_logical_filters(
+                filter1: Dict[str, Any], filter2: Dict[str, Any], logical_op: str
+            ) -> Dict[str, Any]:
+                if filter1["operator"] == filter2["operator"] == logical_op:
+                    return {"operator": logical_op, "conditions": filter1["conditions"] + filter2["conditions"]}
```

Comment:
This doesn’t handle if the conditions contain filters on the same field. This could cause an invalid filter to be made.